x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java

```diff
@@ -73,15 +74,28 @@ protected void onStartJob(long now) {
         String indexName = transformConfig.getDestination();
 
         return pivot.extractResults(agg, getFieldMappings(), getStats()).map(document -> {
+            String id = (String) document.get(DataFrameField.DOCUMENT_ID_FIELD);
+
+            if (id == null) {
+                throw new RuntimeException("Expected a document id but got null.");
+            }
+
             XContentBuilder builder;
             try {
                 builder = jsonBuilder();
-                builder.map(document);
+                builder.startObject();
+                for (Map.Entry<String, ?> value : document.entrySet()) {
+                    // skip all internal fields
+                    if (value.getKey().startsWith("_") == false) {
+                        builder.field(value.getKey(), value.getValue());
+                    }
+                }
+                builder.endObject();
             } catch (IOException e) {
                 throw new UncheckedIOException(e);
             }
 
-            IndexRequest request = new IndexRequest(indexName).source(builder);
+            IndexRequest request = new IndexRequest(indexName).source(builder).id(id);
```

Comment:
I know the probabilities are small, so it may not be worth considering. But, this bugs me a little bit as there is the small chance that two documents will get the same ID and one will overwrite the other.

I suppose as this is written now, there is no way to determine if this is an update or not. Time and scale will tell if this ends up being an intractable problem.

Reply:
In a previous job I worked with an installer that maintained a list of files it had installed by remembering _only_ the MD5 hash of the full installed path.  MD5 is a 128 bit hash function, like murmur hash 128.  In the space of 1 year there were two bugs caused by the same MD5 hash being generated for different paths.  Git uses SHA1 hashes to represent commits.  SHA1 is 160 bits.  I have never known Git mix up two commits in a repo due to generating the same SHA1 for them.  If we use these two (not very scientific) data points then we could surmise that adding an extra 32 bits of entropy to a 128 bit hash turns collisions from "rare but not rare enough" to "a few times in the lifetime of the universe rare".  The algorithm in this PR is already adding 8-40 bits of extra entropy by including the first byte of the first (up to) 5 fields to the final ID.  A computationally very cheap way of further avoiding collisions could be to also append the length of the buffer that gets hashed to the final ID.

Reply:
I definitely defer to you and @hendrikmuhs on this as I have very little experience in this matter. If y'all think the entropy is good, then it probably is :).

Given what we know, there will probably be more bugs introduced trying to implement collision detection/recovery than ones encountered due to ID entropy.